File: app/src/main/java/com/hm/animationdemo/activity/LayoutAnimationActivity.java
package com.hm.animationdemo.activity;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.animation.LayoutAnimationController;
import android.widget.ListView;

import com.hm.animationdemo.R;
import com.hm.animationdemo.adapter.ListViewAdapter;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class LayoutAnimationActivity extends AppCompatActivity {

    @BindView(R.id.list_view)
    ListView listView;

    private List<String> dataList;
    private ListViewAdapter adapter;

    public static void launch(Activity context) {
        Intent starter = new Intent(context, LayoutAnimationActivity.class);
        context.startActivity(starter);
        context.overridePendingTransition(R.anim.enter_ainm, R.anim.exit_ainm);

    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_layout_animation);
        ButterKnife.bind(this);
        Animation animation = AnimationUtils.loadAnimation(this, R.anim.view_animation);
        LayoutAnimationController controller = new LayoutAnimationController(animation);
        controller.setDelay(0.5f);
        controller.setOrder(LayoutAnimationController.ORDER_NORMAL);
        listView.setLayoutAnimation(controller);
        dataList = new ArrayList<>();
        for (int i = 0; i < 50; i++) {
            dataList.add("string" + i);
        }
        adapter = new ListViewAdapter(this, android.R.layout.simple_list_item_1, dataList);
        listView.setAdapter(adapter);
    }

}

File: app/src/main/java/com/hm/animationdemo/activity/LayoutAnimationActivity.kt
package com.hm.animationdemo.activity

import android.animation.LayoutTransition
import android.animation.ObjectAnimator
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.Button
import android.widget.LinearLayout
import androidx.appcompat.app.AppCompatActivity
import com.hm.animationdemo.R
import kotlinx.android.synthetic.main.activity_layout_animation.*

class LayoutAnimationActivity : AppCompatActivity() {

    companion object {


        fun launch(context: Context) {
            val intent = Intent(context, LayoutAnimationActivity::class.java)
            context.startActivity(intent)
        }
    }


    private var leftPosition = 0

    private var rightPosition = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_layout_animation)

        transition1()

        btnAdd.setOnClickListener {

            //addLeftView()
            addRightView()
        }

        btnRemove.setOnClickListener {
            //removeLeftView()
            removeRightView()
        }

    }

    private fun transition1() {
        val transition = LayoutTransition()

        val animIn = ObjectAnimator.ofFloat(null, "rotation", 0f, 1f)
        transition.setAnimator(LayoutTransition.APPEARING, animIn)

        val animOut = ObjectAnimator.ofFloat(null, "rotation", 0f, 90f, 0f)

        transition.setAnimator(LayoutTransition.DISAPPEARING, animOut)
        transition.setDuration(2000)

        llRightContainer.layoutTransition = transition
    }


    private fun addLeftView() {
        val button = Button(this)
        button.text = "button$leftPosition"
        val layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT)
        button.layoutParams = layoutParams
        llLeftContainer.addView(button, leftPosition++)
    }

    private fun removeLeftView() {
        if (leftPosition > 0) {
            llLeftContainer.removeViewAt(0)
            leftPosition--
        }
    }

    private fun addRightView() {
        val button = Button(this)
        button.text = "button$rightPosition"
        val layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT)
        button.layoutParams = layoutParams
        llRightContainer.addView(button, rightPosition++)
    }

    private fun removeRightView() {
        if (rightPosition > 0) {
            llRightContainer.removeViewAt(0)
            rightPosition--
        }
    }

}
